### PR TITLE
Remove unnecessary buffer copy from EventPipe::WriteEvent pipeline

### PR DESCRIPTION
**Pull request by nategraf**
This PR goes about removing an unnecessary buffer copy in the EventPipelineEventProvider by passing the data structures as an array of EventData blobs rather than requiring a flat buffer. To facilitate this alternate structuring in the current code path an encapsulation class, EventPipeEventPayload, is added to the native code

Connected: #13238

### CODE REVIEW

**Comment by nategraf** [comment]
This change will need more testing, and possibly wait until for functional tests are in the CI, to be merged

File: src/vm/eventpipe.cpp [hunk 1]
```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
...
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_size > 0){
+        if (!this->IsFlattened()){
+            m_pData = new BYTE[m_size];
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.cpp [hunk 2]
```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
...
+
+    if(m_size > 0){
+        if(this->IsFlattened()){
+            memcpy(pDst, m_pData, m_size);
+        }
+
+        else if(m_pBlobs != NULL){
+            unsigned int offset = 0;
+            for(int i=0; i<m_blobCount; i++){
+                memcpy(pDst + offset, m_pBlobs, sizeof(EventData));
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.cpp [hunk 3]
```diff
@@ -289,6 +395,34 @@ void EventPipe::WriteEvent(EventPipeEvent &event, BYTE *pData, unsigned int leng
...
         GC_NOTRIGGER;
         MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    EventPipeEventPayload payload(pData, length);
+    EventPipe::WriteEventInternal(event, payload, pActivityId, pRelatedActivityId);
+}
+
+void EventPipe::WriteEventBlob(EventPipeEvent &event, EventData **pBlobs, unsigned int blobCount, LPCGUID pActivityId, LPCGUID pRelatedActivityId)
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.cpp [hunk 4]
```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
...
+    }
+    CONTRACTL_END;
+
+    if(m_performedAllocation){
+        delete[] m_pData;
+        m_pData = NULL;
+    }
+}
+
+void EventPipeEventPayload::Flatten()
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.cpp [hunk 5]
```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
...
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_performedAllocation){
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.cpp [hunk 6]
```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
...
+    m_pData = NULL;
+    m_pBlobs = pBlobs;
+    m_blobCount = blobCount;
+    m_performedAllocation = false;
+
+    S_UINT32 tmp_size = S_UINT32(blobCount) * S_UINT32(sizeof(EventData));
+    if (tmp_size.IsOverflow()){
+        m_pBlobs = NULL;
+        m_blobCount = 0; // ?
+        m_size = 0; // ?
```
Comment (nategraf) [inline src/vm/eventpipe.cpp]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.h [hunk 7]
```diff
@@ -29,6 +29,70 @@ typedef void (*EventPipeCallback)(
...
+private:
+    BYTE *m_pData;
+    EventData **m_pBlobs;
+    unsigned int m_blobCount;
+    unsigned int m_size;
+    bool m_performedAllocation;
+
+public:
+    // Build this payload with a flat buffer inside
+    EventPipeEventPayload(byte *pData, unsigned int length);
```
Comment (nategraf) [inline src/vm/eventpipe.h]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

File: src/vm/eventpipe.h [hunk 8]
```diff
@@ -29,6 +29,70 @@ typedef void (*EventPipeCallback)(
...
+    unsigned long ptr;
+    unsigned int size;
+    unsigned int reserved;
+};
+
+class EventPipeEventPayload
+{
+private:
+    BYTE *m_pData;
+    EventData **m_pBlobs;
```
Comment (nategraf) [inline src/vm/eventpipe.h]: This is consistent with other places in the code

File: src/vm/eventpipe.h [hunk 9]
```diff
@@ -29,6 +29,70 @@ typedef void (*EventPipeCallback)(
     void *FilterData,
     void *CallbackContext);
 
+struct EventData
+{
+public:
+    unsigned long ptr;
```
Comment (nategraf) [inline src/vm/eventpipe.h]: Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587

**Comment by nategraf** [comment]
@dotnet-bot test "Tizen armel Cross Debug Build" "Tizen armel Cross Release Build" please

**Comment by nategraf** [comment]
@dotnet-bot test "Tizen armel Cross Release Build" please

**Comment by nategraf** [comment]
I've finished up a functional test and did a manual investigation of the code paths and output files. Everything seems to be in order. @brianrob would you mind giving this another look over?

File: src/mscorlib/src/System/Diagnostics/Eventing/EventPipe.cs [hunk 10]
```diff
@@ -171,5 +171,9 @@ internal static class EventPipeInternal
         [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
         [SuppressUnmanagedCodeSecurity]
         internal static extern unsafe void WriteEvent(IntPtr eventHandle, uint eventID, void* pData, uint length, Guid* activityId, Guid* relatedActivityId);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern unsafe void WriteEventData(IntPtr eventHandle, uint eventID, EventProvider.EventData** pBlobs, uint blobCount, Guid* activityId, Guid* relatedActivityId);
```
Comment (brianrob) [inline src/mscorlib/src/System/Diagnostics/Eventing/EventPipe.cs]: Nit: Rename from blobs to eventData.

File: src/vm/eventpipe.cpp [hunk 11]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pData = pData;
+    m_pBlobs = NULL;
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Rename "blobs" to eventData for clarity.

File: src/vm/eventpipe.cpp [hunk 12]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+    m_blobCount = 0;
+    m_performedAllocation = false;
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Rename m_performedAllocation to m_allocatedData to make it clear what was allocated.

File: src/vm/eventpipe.cpp [hunk 13]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+    }
+
+    if (tmp_size.IsOverflow())
+    {
+        // If there is an overflow, drop the data and create an empty payload
+        m_pBlobs = NULL;
+        m_blobCount = 0;
+        m_size = 0;
+    }
+    else{
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: { on a new line.

File: src/vm/eventpipe.cpp [hunk 14]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_size > 0)
+    {
+        if (!this->IsFlattened())
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Remove this-> (code style guideline).

File: src/vm/eventpipe.cpp [hunk 15]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+        {
+            BYTE* tmp_pData = new (nothrow) BYTE[m_size];
+            if (tmp_pData != NULL)
+            {
+                m_performedAllocation = true;
+                this->CopyData(tmp_pData);
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Remove this->

File: src/vm/eventpipe.cpp [hunk 16]
```diff
@@ -38,6 +40,145 @@ extern "C" void InitProvidersAndEvents();
...
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if (!this->IsFlattened())
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Remove this->

File: src/vm/eventpipe.cpp [hunk 17]
```diff
@@ -289,6 +430,34 @@ void EventPipe::WriteEvent(EventPipeEvent &event, BYTE *pData, unsigned int leng
...
         GC_NOTRIGGER;
         MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    EventPipeEventPayload payload(pData, length);
+    EventPipe::WriteEventInternal(event, payload, pActivityId, pRelatedActivityId);
+}
+
+void EventPipe::WriteEvent(EventPipeEvent &event, EventData **pBlobs, unsigned int blobCount, LPCGUID pActivityId, LPCGUID pRelatedActivityId)
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Rename blobs to eventData.

File: src/vm/eventpipe.cpp [hunk 18]
```diff
@@ -595,4 +774,22 @@ void QCALLTYPE EventPipeInternal::WriteEvent(
     END_QCALL;
 }
 
+void QCALLTYPE EventPipeInternal::WriteEventData(
+    INT_PTR eventHandle,
+    unsigned int eventID,
+    EventData **pBlobs,
```
Comment (brianrob) [inline src/vm/eventpipe.cpp]: Nit: Rename from blobs to eventData.

File: src/vm/eventpipe.h (hunk 8 above)
Comment (brianrob) [inline src/vm/eventpipe.h]: Nit: Rename m_pBlobs to m_pEventData.

Comment (brianrob) [inline src/vm/eventpipe.h]: And m_blobCount to m_eventDataCount.

**Pull request merged**